### PR TITLE
ref(ts): Improve implementation of CodeTabs

### PR DESCRIPTION
**Pull request by evanpurkhiser**
This component had a number of React anti patterns

### CODE REVIEW

File: src/components/codeTabs.tsx [hunk 1]
```diff
@@ -73,58 +93,78 @@ export function CodeTabs({children}: Props) {
...
   const finalSelection =
-    sharedSelectionChoice || localSelectionChoice || possibleChoices[0];
+    sharedSelectionChoice ?? localSelectionChoice ?? possibleChoices[0];
 
   // Whenever local selection and the final selection are not in sync, the local
   // selection is updated from the final one.  This means that when the shared
   // selection moves to something that is unsupported by the block it stays on
   // its last selection.
-  if (localSelection !== finalSelection) {
-    setLocalSelection(finalSelection);
```
Comment (evanpurkhiser) [inline src/components/codeTabs.tsx:84]: calling set state in a render 😮‍💨

File: src/components/codeTabs.tsx [hunk 2]
```diff
@@ -21,50 +23,68 @@ const LANGUAGES = {
...
+      return title;
+    }
+    if (!language) {
+      return 'Text';
+    }
+    if (language in HUMAN_LANGUAGE_NAMES) {
+      return HUMAN_LANGUAGE_NAMES[language];
+    }
+
+    return language[0].toUpperCase() + language.substring(1);
```
Comment (evanpurkhiser) [inline src/components/codeTabs.tsx:71]: Tried to make this a bit easier to understand

File: src/components/codeTabs.tsx [hunk 3]
```diff
@@ -1,9 +1,11 @@
 import React, {useContext, useEffect, useRef, useState} from 'react';
+import styled from '@emotion/styled';
 
+import {CodeBlockProps} from './codeBlock';
 import {CodeContext} from './codeContext';
 
 // human readable versions of names
-const LANGUAGES = {
+const HUMAN_LANGUAGE_NAMES = {
```
Comment (evanpurkhiser) [inline src/components/codeTabs.tsx:8]: Better name

File: src/components/codeTabs.tsx [hunk 4]
```diff
@@ -20,51 +22,69 @@ const LANGUAGES = {
   yml: 'YAML',
 };
 
-type Props = {
-  children: JSX.Element | JSX.Element[];
-};
+interface CodeTabProps {
+  children: React.ReactElement<CodeBlockProps> | React.ReactElement<CodeBlockProps>[];
+}
```
Comment (evanpurkhiser) [inline src/components/codeTabs.tsx:27]: Correct typing on what the children are, so we know the props

File: src/components/codeTabs.tsx [hunk 5]
```diff
@@ -20,51 +22,69 @@ const LANGUAGES = {
...
-  const tabBarRef = useRef(null);
+  const [localSelection, setLocalSelection] = useState<string | null>(null);
+  const [lastScrollOffset, setLastScrollOffset] = useState<number | null>(null);
+  const containerRef = useRef<HTMLDivElement>(null);
+
+  // When the selection switches we scroll so that the box that was toggled
+  // stays scrolled like it was before. This is because boxes above the changed
+  // box might also toggle and change height.
+  useEffect(() => {
+    if (containerRef.current === null) {
```
Comment (evanpurkhiser) [inline src/components/codeTabs.tsx:46]: This was a type error, since it can be null

File: src/components/codeTabs.tsx [hunk 6]
```diff
@@ -20,51 +22,69 @@ const LANGUAGES = {
...
   // final selection is what is then rendered.
 
   const codeContext = useContext(CodeContext);
-  const [sharedSelection, setSharedSelection] = codeContext.sharedCodeSelection;
-  const [localSelection, setLocalSelection] = useState(null);
-  const [lastScrollOffset, setLastScrollOffset] = useState(null);
-  const tabBarRef = useRef(null);
+  const [localSelection, setLocalSelection] = useState<string | null>(null);
+  const [lastScrollOffset, setLastScrollOffset] = useState<number | null>(null);
+  const containerRef = useRef<HTMLDivElement>(null);
```
Comment (evanpurkhiser) [inline src/components/codeTabs.tsx:40]: Missing types

File: src/components/codeTabs.tsx [hunk 7]
```diff
@@ -20,51 +22,69 @@ const LANGUAGES = {
...
+    }
+
+    return language[0].toUpperCase() + language.substring(1);
   });
 
   // disambiguate duplicates by enumerating them.
-  const tabTitleSeen = {};
-  possibleChoices = possibleChoices.reduce((arr, tabTitle) => {
-    if (possibleChoices.filter(x => x === tabTitle).length > 1) {
-      const num = (tabTitleSeen[tabTitle] = (tabTitleSeen[tabTitle] || 0) + 1);
```
Comment (evanpurkhiser) [inline src/components/codeTabs.tsx:61]: Tried to make this a bit less crazy as well, though `??=` may not be clear to everyone

File: src/components/codeTabs.tsx [hunk 8]
```diff
@@ -20,51 +22,69 @@ const LANGUAGES = {
...
+    if (hasMultiple) {
+      tabTitleSeen[tabTitle] ??= 0;
+      tabTitleSeen[tabTitle] += 1;
+      possibleChoices[index] = `${tabTitle} ${tabTitleSeen[tabTitle]}`;
     }
-    return arr;
-  }, []);
+  });
+
+  const [sharedSelection, setSharedSelection] = codeContext?.sharedCodeSelection ?? [];
```
Comment (evanpurkhiser) [inline src/components/codeTabs.tsx:87]: Can be null

File: src/components/codeTabs.tsx [hunk 9]
```diff
@@ -73,58 +93,78 @@ export function CodeTabs({children}: Props) {
...
-      <div className="tab-bar">{names}</div>
+    <Container ref={containerRef}>
+      <TabBar>{buttons}</TabBar>
       <div className="tab-content">{code}</div>
-    </div>
+    </Container>
   );
 }
+
+const Container = styled('div')`
```
Comment (evanpurkhiser) [inline src/components/codeTabs.tsx:134]: Moved the styles out of scss

**Review by scttcper — APPROVED** [review]

File: src/components/codeTabs.tsx (hunk 7 above)
Comment (shanamatthews) [reply src/components/codeTabs.tsx:61]: Yeah, and operators can be hard to google. You could add a comment with the operator name (Nullish coalescing assignment, I think)

**Review by shanamatthews — APPROVED** [review]

**Pull request merged**